core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java

```diff
@@ -379,141 +295,72 @@ public Throwable cause() {
-            return oldWhenCancelled;
-        }
-    }
-
-    @Override
-    @Deprecated
-    public CompletableFuture<Void> whenTimingOut() {
-        final TimeoutFuture whenTimingOut = this.whenTimingOut;
-        if (whenTimingOut != null) {
-            return whenTimingOut;
-        }
-        final TimeoutFuture timeoutFuture = new TimeoutFuture();
-        if (whenTimingOutUpdater.compareAndSet(this, null, timeoutFuture)) {
-            whenCancelling().thenAccept(cause -> {
-                if (cause instanceof TimeoutException) {
-                    timeoutFuture.doComplete();
-                }
-            });
-            return timeoutFuture;
-        } else {
-            final TimeoutFuture oldWhenTimingOut = this.whenTimingOut;
-            assert oldWhenTimingOut != null;
-            return oldWhenTimingOut;
-        }
+    public boolean isFinished() {
+        return state == State.FINISHED;
     }
 
     @Override
-    @Deprecated
-    public CompletableFuture<Void> whenTimedOut() {
-        final TimeoutFuture whenTimedOut = this.whenTimedOut;
-        if (whenTimedOut != null) {
-            return whenTimedOut;
-        }
-        final TimeoutFuture timeoutFuture = new TimeoutFuture();
-        if (whenTimedOutUpdater.compareAndSet(this, null, timeoutFuture)) {
-            whenCancelled().thenAccept(cause -> {
-                if (cause instanceof TimeoutException) {
-                    timeoutFuture.doComplete();
-                }
-            });
-            return timeoutFuture;
-        } else {
-            final TimeoutFuture oldWhenTimedOut = this.whenTimedOut;
-            assert oldWhenTimedOut != null;
-            return oldWhenTimedOut;
+    public void updateTask(CancellationTask task) {
+        lock.lock();
+        try {
+            if (state != State.FINISHED) {
+                // if the task hasn't been run yet
+                this.task = task;
+                return;
+            }
+        } finally {
+            lock.unlock();
         }
-    }
 
-    private boolean isInitialized() {
-        return pendingTask == noopPendingTask && eventLoop != null;
-    }
-
-    private void addPendingTask(Runnable pendingTask) {
-        if (!pendingTaskUpdater.compareAndSet(this, null, pendingTask)) {
-            for (;;) {
-                final Runnable oldPendingTask = this.pendingTask;
-                assert oldPendingTask != null;
-                if (oldPendingTask == noopPendingTask) {
-                    assert eventLoop != null;
-                    eventLoop.execute(pendingTask);
-                    break;
-                }
-                final Runnable newPendingTask = () -> {
-                    oldPendingTask.run();
-                    pendingTask.run();
-                };
-                if (pendingTaskUpdater.compareAndSet(this, oldPendingTask, newPendingTask)) {
-                    break;
-                }
+        whenCancelled().thenAccept(cause -> {
```

Comment:
This callback is called for all updating tasks.
How about adding this callback only once and use `this.task` instead of `task` in the callback?

Reply:
I didn't really understand your comment here.

> This callback is called for all updating tasks.

Technically, this code path will be called when `updateTask` is invoked after the `CancellationScheduler` is in the `FINISHED` state. The initial timed out task won't be invoked here

>  How about adding this callback only once and use this.task instead of task in the callback?

Are you suggesting that we modify `this.task`? My concern is this will create a race condition with `invokeTask()`. Do you mind adding a suggestion block so I can understand your intention?

Reply:
Oops, my bad. I didn't notice the `return` statement in the above `if` condition so I thought `thenAccept` will be attached for all `task`s. 😓